### PR TITLE
support web app deployment slot -part2

### PR DESCRIPTION
**Pull request by SummerSun**
* Implement IDeployTargetAdapter interface with DeploymentSlot

* Decide the deploy target from pom configuration

* Webapp must exists when deploy to slot, create new slot if not exits

* Update existing unit tests

* DeploymentSlotHandlerImpl unit tests, integrations tests will be included in next change list

### CODE REVIEW

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandlerImpl.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        this.mojo.getLog().info(CREATE_DEPLOYMENT_SLOT);
+
+        createDeploymentSlotWithConfigurationSource(app, slotName, slotSetting.getConfigurationSource());
+
+        this.mojo.getLog().info(CREATE_DEPLOYMENT_SLOT_DONE);
+    }
+
+    protected void createDeploymentSlotWithConfigurationSource (final WebApp app, final String slotName,
+                                                              final String configurationSource) {
+        final DeploymentSlot cSlot = this.mojo.getDeploymentSlot(app, configurationSource);
```
Comment (akaroml) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandlerImpl.java]: what does cSlot mean?

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandler.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+ * license information.
+ */
+
+package com.microsoft.azure.maven.webapp.handlers;
+
+import com.microsoft.azure.maven.auth.AzureAuthFailureException;
+import org.apache.maven.plugin.MojoExecutionException;
+
+public interface DeploymentSlotHandler {
+    void handleDeploymentSlot() throws AzureAuthFailureException, MojoExecutionException;
```
Comment (akaroml) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandler.java]: what does "handle" indicate?
in the implementation, i only see create if not exists operation. please be explicit if there are no other possibilities.

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandlerImpl.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+package com.microsoft.azure.maven.webapp.handlers;
+
+import com.microsoft.azure.management.appservice.DeploymentSlot;
+import com.microsoft.azure.management.appservice.WebApp;
+import com.microsoft.azure.maven.auth.AzureAuthFailureException;
+import com.microsoft.azure.maven.webapp.AbstractWebAppMojo;
+import com.microsoft.azure.maven.webapp.configuration.DeploymentSlotSetting;
+import org.apache.maven.plugin.MojoExecutionException;
+
+public class DeploymentSlotHandlerImpl implements DeploymentSlotHandler {
```
Comment (akaroml) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandlerImpl.java]: will there be other handlers other than this?

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandlerImpl.java (hunk 1 above)
Comment (SummerSun) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandlerImpl.java]: will rename to ```copyConfigurationFromSlot```

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandlerImpl.java (hunk 3 above)
Comment (SummerSun) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandlerImpl.java]: currently no, and this makes it more like a Util class.

Comment (SummerSun) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandlerImpl.java]: Rethink about it, make DeploymentSlotUtils to handle deployment get/check/create functionalities.

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java [hunk 4]
```diff
@@ -348,6 +363,21 @@ public WebApp getWebApp() throws AzureAuthFailureException {
         return null;
     }
 
+    public DeploymentSlot getDeploymentSlot(final WebApp app, final String slotName) {
+        DeploymentSlot slot = null;
+        if (!StringUtils.isEmpty(slotName)) {
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java]: Does StringUtils has `isNotEmpty()` method?

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java [hunk 5]
```diff
@@ -348,6 +363,21 @@ public WebApp getWebApp() throws AzureAuthFailureException {
         return null;
     }
 
+    public DeploymentSlot getDeploymentSlot(final WebApp app, final String slotName) {
+        DeploymentSlot slot = null;
+        if (!StringUtils.isEmpty(slotName)) {
+            try {
+                slot = app.deploymentSlots().getByName(slotName);
+            } catch (NoSuchElementException deploymentSlotNotExistException) {
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java]: swallow the deploymentSlotNotExistException is fine here. But we need to differentiate:
- user doesn't set these configuration
- user set but the slot name is invalid

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/DeploymentSlotSetting.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for
+ * license information.
+ */
+
+package com.microsoft.azure.maven.webapp.configuration;
+
+/**
+ * Deployment SLot setting class.
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/DeploymentSlotSetting.java]: SLot -> Slot

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java (hunk 4 above)
Comment (SummerSun) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java]: Actually there is ^_^

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandler.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+public class DeploymentSlotHandler {
+    private static final String INVALID_SLOT_SETTINGS = "<deploymentSlotSetting> is NULL." +
+            "Need configure it in pom.xml for deploying it to deployment slot.";
+    private static final String EMPTY_CONFIGURATION_SOURCE =
+            "Empty configuration source, create a brand new deployment slot without any configuration.";
+    private static final String DEFAULT_CONFIGURATION_SOURCE =
+            "Null or unrecognized configuration source, create deployment slot and copy configuration from parent.";
+    private static final String CREATE_DEPLOYMENT_SLOT = "Start create deployment slot...";
+    private static final String CREATE_DEPLOYMENT_SLOT_DONE = "Successfully created deployment slot.";
+    private static final String INVALID_SLOT_NAME = "<slotName> is invalid for creating deployment slot.";
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandler.java]: Can be more specific here. Here `invalid` actually means empty, right?

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandler.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        this.mojo = mojo;
+    }
+
+    public void createDeploymentSlotIfNotExist() throws MojoExecutionException, AzureAuthFailureException {
+        final DeploymentSlotSetting slotSetting = this.mojo.getDeploymentSlotSetting();
+        assureValidSlotSetting(slotSetting);
+
+        final WebApp app = this.mojo.getWebApp();
+        final String slotName = slotSetting.getSlotName();
+        final DeploymentSlot slot = this.mojo.getDeploymentSlot(app, slotName);
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandler.java]: if `slot` is only used in the if block, then we can remove this temp variable for now.

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandler.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                                        final String configurationSource) throws MojoExecutionException {
+        assureValidSlotName(slotName);
+        final DeploymentSlot configurationSourceSlot = this.mojo.getDeploymentSlot(app, configurationSource);
+        final DeploymentSlot.DefinitionStages.Blank definedSlot = app.deploymentSlots().define(slotName);
+
+        this.mojo.getLog().info(CREATE_DEPLOYMENT_SLOT);
+
+        if (configurationSourceSlot != null) {
+            definedSlot.withConfigurationFromDeploymentSlot(configurationSourceSlot).create();
+        } else if ("".equalsIgnoreCase(configurationSource)) {
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandler.java]: `""` is also not null, so this block will not be executed actually. The if block here is buggy.

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/ConfigurationSourceType.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+ * license information.
+ */
+
+package com.microsoft.azure.maven.webapp.configuration;
+
+import org.codehaus.plexus.util.StringUtils;
+
+import java.util.Locale;
+
+public enum ConfigurationSourceType {
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/configuration/ConfigurationSourceType.java]: Let's put some comments here to describe what's the meaning of each Enum

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandler.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    protected void assureValidSlotSetting(final DeploymentSlotSetting slotSetting) throws MojoExecutionException {
+        if (slotSetting == null) {
+            throw new MojoExecutionException(INVALID_SLOT_SETTINGS);
+        }
+    }
+
+    protected void assureValidSlotName(final String slotName) throws MojoExecutionException {
+        final Pattern pattern = Pattern.compile(SLOT_NAME_PATTERN, Pattern.CASE_INSENSITIVE);
+
+        if (!pattern.matcher(slotName).matches()) {
```
Comment (jdneo) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/handlers/DeploymentSlotHandler.java]: Check NPE?

**Pull request merged**